game.js

```diff
@@ -624,106 +615,97 @@ module.exports = function createGame(options) {
-            if (state.state.name != stateNames.REVEAL_INFLUENCE) {
-                throw new GameException('Incorrect state');
+            setState({
+                name: stateNames.ACTION_RESPONSE,
+                playerIdx: playerIdx,
+                action: command.action,
+                target: command.target,
+                message: message
+            });
+            resetAllows(playerIdx);
+        }
+    ];
+    behaviors[['challenge', stateNames.ACTION_RESPONSE]] = [
+        playerIsUndead,
+        invalidPlayer('playerIdx'),
+        (command, playerIdx) => {
+            const action = actions[state.state.action];
+            if (!action) {
+                throw new GameException('Unknown action');
+            }
+            if (!action.roles) {
+                throw new GameException('Action cannot be challenged');
             }
+            challenge(playerIdx, state.state.playerIdx, getGameRole(action.roles));
+        }
+    ];
+    behaviors[['challenge', stateNames.BLOCK_RESPONSE]] = [
+        playerIsUndead,
+        invalidPlayer('target'),
+        (command, playerIdx) => {
+            challenge(playerIdx, state.state.target, state.state.blockingRole);
+        }
+    ];
+    behaviors[['reveal', stateNames.REVEAL_INFLUENCE]] = [
+        (command, playerIdx) => {
             if (state.state.playerToReveal != playerIdx) {
                 throw new GameException('Not your turn to reveal an influence');
             }
-            for (i = 0; i < playerState.influence.length; i++) {
-                var influence = playerState.influence[i];
-                if (influence.role == command.role && !influence.revealed) {
-                    influence.revealed = true;
-                    playerState.influenceCount--;
-                    addHistory(state.state.reason, curTurnHistGroup(), '%s; {%d} revealed %s', state.state.message, playerIdx, command.role);
-                    if (state.state.reason == 'incorrect-challenge') {
-                        if (afterIncorrectChallenge()) {
-                            nextTurn();
-                        }
-                    } else if (state.state.reason == 'successful-challenge') {
-                        if (afterSuccessfulChallenge()) {
-                            nextTurn();
-                        }
-                    } else {
-                        // The reveal is due to a coup or assassination.
+            for (var i = 0; i < state.players[playerIdx].influence.length; i++) {
+                var influence = state.players[playerIdx].influence[i];
+                if (influence.role != command.role || influence.revealed) {
+                    continue
+                }
+                influence.revealed = true;
+                state.players[playerIdx].influenceCount--;
+                addHistory(state.state.reason, curTurnHistGroup(), '%s; {%d} revealed %s', state.state.message, playerIdx, command.role);
+                const reasonMap = {
```

Comment:
Is this really better than an `if` statement??

Reply:
I think I overdid it here lol